File: tokenrazor/core/reporter.py
"""剪枝报告格式化输出。包含 Token 统计、费用节省一目了然。"""

from typing import Optional

from .pruner import PruneResult
from .pricing import calculate_cost, cost_report, format_cost


class Report:
    """生成人类可读的剪枝报告。"""

    BAR_WIDTH = 40

    @classmethod
    def text(cls, result: PruneResult, show_diff: bool = False,
             model: Optional[str] = None) -> str:
        """生成文本格式报告，包含费用节省。"""
        lines = []
        lines.append("╔════════════════════════════════════════╗")
        lines.append("║        TokenRazor 剪枝报告            ║")
        lines.append("╚════════════════════════════════════════╝")
        lines.append("")

        s = result.stats
        ratio = result.compression_ratio

        # Token 统计
        lines.append(f"  📊 Token 统计")
        lines.append(f"  ┌─────────────────────────────────────┐")
        lines.append(f"  │ 原始 Tokens    {s['original_tokens']:>10}       │")
        lines.append(f"  │ 剪后 Tokens    {s['pruned_tokens']:>10}       │")
        lines.append(f"  │ 节约 Tokens    {s['saved_tokens']:>10}       │")
        lines.append(f"  │ 压缩率         {s['saved_percent']:>9}%       │")
        lines.append(f"  └─────────────────────────────────────┘")
        lines.append("")

        # 可视化进度条
        saved_bar = int(cls.BAR_WIDTH * (1 - ratio))
        used_bar = cls.BAR_WIDTH - saved_bar
        bar = "█" * used_bar + "░" * saved_bar
        lines.append(f"  [{bar}] {s['saved_percent']}% 压缩")
        lines.append("")

        # 费用节省（商业价值核心）
        cost = cost_report(s['original_tokens'], s['pruned_tokens'], model=model)
        lines.append(f"  💰 费用节省（{cost['model']} @ ${cost['price_per_1m']}/1M tokens）")
        lines.append(f"  ┌─────────────────────────────────────┐")
        lines.append(f"  │ 原始费用      {format_cost(cost['original_cost']):>10}        │")
        lines.append(f"  │ 剪后费用      {format_cost(cost['pruned_cost']):>10}        │")
        lines.append(f"  │ 本次节省      {format_cost(cost['saved_cost']):>10}        │")
        lines.append(f"  └─────────────────────────────────────┘")
        lines.append("")

        # 冗余详情
        if result.removed_spans:
            lines.append(f"  移除冗余段: {len(result.removed_spans)} 处")
            for m in result.removed_spans:
                lines.append(f"    · {m.reason:<12} (置信度 {m.confidence:.0%})")
        else:
            lines.append("  未发现显著冗余。")
        lines.append("")

        if show_diff:
            lines.append("─" * 50)
            lines.append("  【原始文本】")
            lines.append(result.original[:500])
            lines.append("")
            lines.append("  【剪后文本】")
            lines.append(result.pruned[:500])
            lines.append("")

        return "\n".join(lines)

    @classmethod
    def markdown(cls, result: PruneResult, model: Optional[str] = None) -> str:
        """生成 Markdown 格式报告。"""
        s = result.stats
        cost = cost_report(s['original_tokens'], s['pruned_tokens'], model=model)
        saved = format_cost(cost['saved_cost'])
        lines = [
            "## TokenRazor 剪枝报告",
            "",
            "| 指标 | 值 |",
            "|---|---|",
            f"| 原始 Tokens | {s['original_tokens']} |",
            f"| 剪后 Tokens | {s['pruned_tokens']} |",
            f"| 节约 Tokens | {s['saved_tokens']} |",
            f"| 压缩率 | {s['saved_percent']}% |",
            f"| 节省费用 | {saved} |",
            "",
            f"共移除 {len(result.removed_spans)} 处冗余。",
            "",
        ]
        return "\n".join(lines)

    @classmethod
    def json(cls, result: PruneResult, model: Optional[str] = None) -> dict:
        """生成 JSON 格式报告，包含费用数据。"""
        cost = cost_report(
            result.stats['original_tokens'],
            result.stats['pruned_tokens'],
            model=model,
        )
        return {
            "compression": result.stats,
            "cost": cost,
            "removed_spans": [
                {
                    "reason": m.reason,
                    "confidence": m.confidence,
                    "start": m.start,
                    "end": m.end,
                }
                for m in result.removed_spans
            ],
            "pruned_text": result.pruned,
            "original_text": result.original,
        }

File: rust/src/reporter.rs
/// 报告生成 — 文本和 JSON 格式

use serde::Serialize;

use crate::pruner::PruneResult;
use crate::pricing;

/// 剪枝统计
#[derive(Debug, Clone, Serialize)]
pub struct PruneStats {
    pub original_tokens: u64,
    pub pruned_tokens: u64,
    pub saved_tokens: u64,
    pub saved_percent: f64,
}

/// 费用信息
#[derive(Debug, Clone, Serialize)]
pub struct CostInfo {
    pub original_cost: f64,
    pub pruned_cost: f64,
    pub saved_cost: f64,
    pub saved_percent: f64,
    pub model: String,
    pub price_per_1m: f64,
    pub currency: String,
}

/// JSON 报告
#[derive(Debug, Clone, Serialize)]
pub struct JsonReport {
    pub compression: PruneStats,
    pub cost: CostInfo,
    pub removed_spans: Vec<crate::scanner::RedundancyMatch>,
    pub pruned_text: String,
    pub original_text: String,
}

/// 生成 JSON 格式报告
pub fn json_report(result: &PruneResult, model: &str) -> JsonReport {
    let cost = pricing::cost_report(result.original_tokens, result.pruned_tokens, model);

    JsonReport {
        compression: PruneStats {
            original_tokens: result.original_tokens,
            pruned_tokens: result.pruned_tokens,
            saved_tokens: result.saved_tokens(),
            saved_percent: result.saved_percent(),
        },
        cost: CostInfo {
            original_cost: cost.original_cost,
            pruned_cost: cost.pruned_cost,
            saved_cost: cost.saved_cost,
            saved_percent: cost.saved_percent,
            model: cost.model,
            price_per_1m: cost.price_per_1m,
            currency: cost.currency,
        },
        removed_spans: result.removed_spans.clone(),
        pruned_text: result.pruned.clone(),
        original_text: result.original.clone(),
    }
}

/// 生成文本格式报告
pub fn text_report(result: &PruneResult, model: &str) -> String {
    let cost = pricing::cost_report(result.original_tokens, result.pruned_tokens, model);
    let bar = filled_bar(result.saved_percent());

    let mut lines = Vec::new();
    lines.push("╔════════════════════════════════════════╗".to_string());
    lines.push("║        TokenRazor 剪枝报告            ║".to_string());
    lines.push("╚════════════════════════════════════════╝".to_string());
    lines.push(String::new());

    // Token 统计
    lines.push("  📊 Token 统计".to_string());
    lines.push("  ┌─────────────────────────────────────┐".to_string());
    lines.push(format!("  │ 原始 Tokens  {:>8}       │", result.original_tokens));
    lines.push(format!("  │ 剪后 Tokens  {:>8}       │", result.pruned_tokens));
    lines.push(format!("  │ 节约 Tokens  {:>8}       │", result.saved_tokens()));
    lines.push(format!("  │ 压缩率      {:>7.1}%       │", result.saved_percent()));
    lines.push("  └─────────────────────────────────────┘".to_string());
    lines.push(String::new());

    // 进度条
    lines.push(format!("  [{}] {:.1}% 压缩", bar, result.saved_percent()));
    lines.push(String::new());

    // 费用节省
    lines.push(format!("  💰 费用节省（{} @ ${:.2}/1M tokens）",
        cost.model, cost.price_per_1m));
    lines.push("  ┌─────────────────────────────────────┐".to_string());
    lines.push(format!("  │ 原始费用     {:>12}        │", pricing::format_cost(cost.original_cost)));
    lines.push(format!("  │ 剪后费用     {:>12}        │", pricing::format_cost(cost.pruned_cost)));
    lines.push(format!("  │ 本次节省     {:>12}        │", pricing::format_cost(cost.saved_cost)));
    lines.push("  └─────────────────────────────────────┘".to_string());
    lines.push(String::new());

    // 移除统计
    if !result.removed_spans.is_empty() {
        use std::collections::HashMap;
        let mut reasons: HashMap<&str, u32> = HashMap::new();
        for s in &result.removed_spans {
            *reasons.entry(&s.reason).or_insert(0) += 1;
        }
        lines.push(format!("  移除冗余段: {} 处", result.removed_spans.len()));
        for (reason, _count) in &reasons {
            let conf = &result.removed_spans.iter()
                .find(|s| s.reason == *reason)
                .map(|s| s.confidence)
                .unwrap_or(0.0);
            lines.push(format!("    · {:<15} (置信度 {:>2}%)", reason, (conf * 100.0) as u32));
        }
    } else {
        lines.push("  未发现显著冗余。".to_string());
    }
    lines.push(String::new());

    lines.join("\n")
}

/// 填充进度条
fn filled_bar(percent: f64) -> String {
    let width = 40;
    let filled = ((percent / 100.0) * width as f64).round() as usize;
    let filled = filled.min(width);
    let empty = width - filled;
    format!("{}░{}",
        "█".repeat(filled),
        "░".repeat(empty))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::pruner::Pruner;

    #[test]
    fn test_text_report_basic() {
        let p = Pruner::new(&["filler"]);
        let r = p.prune("让我想想。答案：42");
        let report = text_report(&r, "gpt-4o");
        assert!(report.contains("TokenRazor"));
        assert!(report.contains("Token"));
    }

    #[test]
    fn test_json_report_structure() {
        let p = Pruner::new(&["filler"]);
        let r = p.prune("让我想想。答案：42");
        let j = json_report(&r, "gpt-4o");
        assert!(j.compression.saved_tokens > 0 || j.compression.original_tokens > 0);
        assert_eq!(j.cost.model, "gpt-4o");
    }

    #[test]
    fn test_bar_100_percent() {
        let bar = filled_bar(100.0);
        assert!(bar.contains('█'));
        assert!(!bar.starts_with('░'));
    }

    #[test]
    fn test_bar_0_percent() {
        let bar = filled_bar(0.0);
        assert!(bar.contains('░'));
        assert!(!bar.contains('█'));
    }
}
